File: js/bench/datasets.ts
// The pure dataset-generation layer, shared by every bench instrument in this
// directory: the cardinality-controlled generator the comparative suite and
// the dict-memory instrument drive, the query-pattern probes derived from it,
// and the dense-cube datasets the instrumented CodSpeed suite uses for its
// routing-pattern tasks.
//
// PURITY CONTRACT: this module may import only rdf-data-factory and types.
// No store library (oxigraph, rdf-stores, or the Vortex wasm module) may be
// reachable from here — codspeed.bench.ts imports this file and runs under
// CodSpeed's Valgrind instrumentation, where loading a foreign multi-MB wasm
// module would pollute every measurement (see util.ts, the other pure
// module). The impure adapter/measurement half lives in shared.ts, which
// re-exports this module so the multi-process workers keep a single import.

import { DataFactory } from 'rdf-data-factory';
import type { Quad, Term } from '@rdfjs/types';

export const df = new DataFactory();

// ─── Cardinality-controlled dataset ──────────────────────────────────────────
//
// Term cardinality is an explicit knob, independent of row count. The previous
// generator drew D³ rows from a namespace of only D IRIs (128 distinct terms for
// 2,097,152 triples), which made every store's term handling — dictionaries,
// interning, string storage — invisible to the benchmark, and is nothing like
// real RDF, where distinct terms scale with the data.
//
// Uniqueness: term indices are `i % k` per role, so quad i maps to the residue
// tuple (i mod nSubj, i mod nPred, i mod nObj, i mod nGraph). By the Chinese
// Remainder Theorem that map is injective over `i < lcm(...)`, so making the
// four moduli pairwise coprime (their lcm is then their product) and checking
// the product covers `n` guarantees every generated quad is distinct — with no
// dedupe set, which at these row counts would itself dominate memory.
//
// Deliberate consequence: index 0 exists for every role, so row 0 satisfies all
// four single-term probes and every pattern in `datasetProbes` matches at least
// one row.

const BASE = 'http://data.example.org';

export interface DatasetOpts {
    /** distinct subjects / n (default 0.1 — see `SUBJECT_RATIO_DEFAULT`). */
    subjectRatio?: number;
    /** distinct predicates: a small closed vocabulary, as in real data (default 32). */
    predicates?: number;
    /** distinct objects / n (default 0.5). */
    objectRatio?: number;
    /** fraction of distinct objects that are literals rather than IRIs (default 0.4). */
    literalFrac?: number;
    /** distinct named graphs; 1 means the default graph only (default 1). */
    graphs?: number;
}

/** Distinct subjects per row, i.e. the reciprocal of how many triples describe
 *  each resource. 0.1 is ten triples per subject.
 *
 *  This is the knob that decides how much of the dataset is *terms*: the
 *  dictionary holds `subjectRatio + objectRatio` terms per quad, so the two
 *  ratios alone say whether there are more distinct terms than rows.
 *
 *  It was 1.0 — a distinct subject for every row. That is not how RDF looks: a
 *  resource is normally described by several properties, so its subject recurs
 *  across that many triples. 1.0 also made `S` match exactly one row, and with
 *  `PO`/`SPO`/`SPOG` matching one apiece, five of the seven probes returned two
 *  rows or fewer — so the comparison measured query *setup* almost to the
 *  exclusion of decoding, on a dictionary half again the size of the dataset.
 *
 *  Set `BENCH_SUBJ_RATIO=1.0` to get that back deliberately: it is a reasonable
 *  dictionary-stress configuration, and `bench:dict-memory` asks for it
 *  explicitly for exactly that reason. It is a poor default. */
export const SUBJECT_RATIO_DEFAULT = 0.1;

/** Env-overridable defaults, so a sweep can vary cardinality without new code. */
export function datasetOpts(o: DatasetOpts = {}): Required<DatasetOpts> {
    return {
        subjectRatio:
            o.subjectRatio ?? Number(process.env.BENCH_SUBJ_RATIO ?? SUBJECT_RATIO_DEFAULT),
        predicates: o.predicates ?? Number(process.env.BENCH_PREDICATES ?? 32),
        objectRatio: o.objectRatio ?? Number(process.env.BENCH_OBJ_RATIO ?? 0.5),
        literalFrac: o.literalFrac ?? Number(process.env.BENCH_LITERAL_FRAC ?? 0.4),
        graphs: o.graphs ?? Number(process.env.BENCH_GRAPHS ?? 1),
    };
}

function gcd(a: number, b: number): number {
    while (b) [a, b] = [b, a % b];
    return a;
}

/** Per-role term counts: the requested values nudged up until pairwise coprime. */
export function moduli(n: number, opts: DatasetOpts = {}): {
    nSubj: number; nPred: number; nObj: number; nGraph: number; terms: number;
} {
    const o = datasetOpts(opts);
    const want = [
        Math.max(1, Math.round(n * o.subjectRatio)),
        Math.max(1, Math.round(o.predicates)),
        Math.max(1, Math.round(n * o.objectRatio)),
        Math.max(1, Math.round(o.graphs)),
    ];
    const got: number[] = [];
    for (const w of want) {
        let k = w;
        while (got.some((g) => gcd(g, k) !== 1)) k++;
        got.push(k);
    }
    const [nSubj, nPred, nObj, nGraph] = got;
    // Product == lcm because they are pairwise coprime; see the CRT note above.
    // Use logs: the product overflows the float mantissa long before it matters.
    const logProduct = got.reduce((acc, k) => acc + Math.log(k), 0);
    if (logProduct < Math.log(n)) {
        throw new Error(
            `dataset cardinality too low for ${n} distinct quads: ` +
            `${nSubj}x${nPred}x${nObj}x${nGraph} cannot cover it — raise a ratio`,
        );
    }
    // The default graph is a term too, and it is not one of the named graphs.
    const terms = nSubj + nPred + nObj + (nGraph === 1 ? 1 : nGraph);
    return { nSubj, nPred, nObj, nGraph, terms };
}

export const subjectTerm = (i: number) =>
    df.namedNode(`${BASE}/resource/2026/subject/${String(i).padStart(9, '0')}`);
export const predicateTerm = (i: number) =>
    df.namedNode(`${BASE}/ontology/2026/property/${String(i).padStart(4, '0')}`);
export const graphTerm = (i: number) =>
    df.namedNode(`${BASE}/graph/2026/named/${String(i).padStart(6, '0')}`);

/** Objects alternate IRI/literal deterministically, in `literalFrac` proportion. */
export function objectTerm(i: number, literalFrac: number): Term {
    return i % 10 < Math.round(literalFrac * 10)
        ? df.literal(`descriptive object value number ${String(i).padStart(9, '0')}`)
        : df.namedNode(`${BASE}/resource/2026/object/${String(i).padStart(9, '0')}`);
}

/** Quad `i` of the `n`-row dataset. Pure in `i`, so probes and prefixes can be
 *  derived without materializing the array. */
export function quadAt(i: number, m: ReturnType<typeof moduli>, literalFrac: number): Quad {
    return df.quad(
        subjectTerm(i % m.nSubj),
        predicateTerm(i % m.nPred),
        objectTerm(i % m.nObj, literalFrac) as never,
        m.nGraph === 1 ? df.defaultGraph() : graphTerm(i % m.nGraph),
    );
}

/** `n` distinct quads whose term cardinality follows `opts`. */
export function genDataset(n: number, opts: DatasetOpts = {}): Quad[] {
    const o = datasetOpts(opts);
    const m = moduli(n, opts);
    const out: Quad[] = new Array(n);
    for (let i = 0; i < n; i++) out[i] = quadAt(i, m, o.literalFrac);
    return out;
}

/** The first `take` quads `genDataset(n, opts)` would produce, without building
 *  the full array — the mutation worker needs only a MUT_BATCH-sized slice for
 *  the delete phase, and `n` can be a couple of million. */
export function genDatasetPrefix(n: number, take: number, opts: DatasetOpts = {}): Quad[] {
    const o = datasetOpts(opts);
    const m = moduli(n, opts);
    const k = Math.min(take, n);
    const out: Quad[] = new Array(k);
    for (let i = 0; i < k; i++) out[i] = quadAt(i, m, o.literalFrac);
    return out;
}

/** A batch of fresh quads in a disjoint namespace, for the add phase. */
export function genFresh(n: number): Quad[] {
    const out: Quad[] = new Array(n);
    for (let i = 0; i < n; i++) {
        out[i] = df.quad(
            df.namedNode(`${BASE}/fresh/2026/subject/${String(i).padStart(9, '0')}`),
            df.namedNode(`${BASE}/fresh/2026/property/0000`),
            df.namedNode(`${BASE}/fresh/2026/object/${String(i).padStart(9, '0')}`),
        );
    }
    return out;
}

// ─── Query patterns (probe terms fixed at index 0, so they always hit rows) ──
export type Pat = { name: string; s: Term | null; p: Term | null; o: Term | null; g: Term | null };
// 'full' (every variable unbound) is measured separately from the selective patterns,
// under FULL_SCAN_OPTS's much lower repetition count (see shared.ts): repeating a
// full-table materialization (all D³ rows, ~2M at the default D=128) ~15x (QUERY_OPTS's
// 5 warmup + 10 timed) reproducibly trips an internal `unreachable` trap in oxigraph's
// wasm build around the 5th repetition on the same store — confirmed in isolation with
// nothing else in the process, so it's not a cross-adapter memory issue. No other
// adapter under test showed any problem with that many repetitions, but a full-table
// dump repeated many times is a heavy op for any store (closer in spirit to `ingest`
// than to a selective query), so the lower budget applies to every adapter uniformly
// rather than singling out the one library that happens to crash on it.
export const FULL_SCAN_PATTERN: Pat = { name: 'full', s: null, p: null, o: null, g: null };

/**
 * Probes for an `n`-row dataset built with `opts`, all bound to term index 0 so
 * every one matches row 0 at minimum — no pattern can silently measure a
 * zero-row query.
 *
 * Selectivity now follows the data rather than the old shared namespace: at the
 * defaults `S` matches ~1 row (subjects are near-unique), `P` ~n/32, `O` ~2, and
 * the conjunctions narrow to a handful. That spread is what real queries look
 * like, but it is *not* comparable to the pre-cardinality numbers — the workers
 * record each pattern's matched-row count alongside its timing so the figures
 * stay interpretable.
 */
export function datasetProbes(n: number, opts: DatasetOpts = {}): {
    triples: Pat[]; quads: Pat[]; full: Pat;
} {
    const o = datasetOpts(opts);
    const m = moduli(n, opts);
    const s0 = subjectTerm(0);
    const p0 = predicateTerm(0);
    const o0 = objectTerm(0, o.literalFrac);
    const g0 = m.nGraph === 1 ? df.defaultGraph() : graphTerm(0);
    return {
        triples: [
            { name: 'S', s: s0, p: null, o: null, g: null },
            { name: 'P', s: null, p: p0, o: null, g: null },
            { name: 'O', s: null, p: null, o: o0, g: null },
            { name: 'PO', s: null, p: p0, o: o0, g: null },
            { name: 'SPO', s: s0, p: p0, o: o0, g: null },
        ],
        quads: [
            { name: 'G', s: null, p: null, o: null, g: g0 },
            { name: 'SPOG', s: s0, p: p0, o: o0, g: g0 },
        ],
        full: FULL_SCAN_PATTERN,
    };
}

// ─── Dense-cube datasets (the instrumented suite's routing-pattern data) ─────
//
// A d³/d⁴ cube over a tiny closed namespace: every routing decision is
// exercised while the dataset stays small enough for Valgrind. Deliberately
// NOT cardinality-realistic — 3·d distinct terms over d³ rows is exactly the
// pathology the generator above exists to avoid — so term-handling-sensitive
// tasks must use `genDataset` instead; the cube is for tasks where routing,
// not term volume, is the object of measurement.

export const EX = 'http://example.org/#';
export const nn = (n: number | string) => df.namedNode(EX + n);

/** d³ triples in the default graph: quad(ex#s, ex#p, ex#o). */
export function genTriples(d: number): Quad[] {
    const out: Quad[] = [];
    for (let s = 0; s < d; s++)
        for (let p = 0; p < d; p++)
            for (let o = 0; o < d; o++) out.push(df.quad(nn(s), nn(p), nn(o)));
    return out;
}

/** d⁴ quads across named graphs: quad(ex#s, ex#p, ex#o, ex#g). */
export function genQuads(d: number): Quad[] {
    const out: Quad[] = [];
    for (let s = 0; s < d; s++)
        for (let p = 0; p < d; p++)
            for (let o = 0; o < d; o++)
                for (let g = 0; g < d; g++) out.push(df.quad(nn(s), nn(p), nn(o), nn(g)));
    return out;
}

/** d³ triples whose objects are literals — plain, language-tagged, typed,
 * and escape-bearing (quotes, backslashes, newlines, and `"@`/`^^` lookalikes
 * inside the value). The other dense-cube datasets are named-node-only, so
 * the instrumented tasks over this one are the only dense-cube coverage of
 * literal escaping on ingest and the unescape path on decode/export. */
export function genLiteralTriples(d: number): Quad[] {
    const dt = df.namedNode('http://www.w3.org/2001/XMLSchema#integer');
    const out: Quad[] = [];
    let i = 0;
    for (let s = 0; s < d; s++)
        for (let p = 0; p < d; p++)
            for (let o = 0; o < d; o++) {
                const object =
                    i % 4 === 0 ? df.literal(`plain value ${o}`)
                    : i % 4 === 1 ? df.literal(`tagged value ${o}`, 'en')
                    : i % 4 === 2 ? df.literal(String(o), dt)
                    : df.literal(`say "hi"@home ^^ line\nbreak back\\slash ${o}`);
                out.push(df.quad(nn(s), nn(p), object));
                i++;
            }
    return out;
}

/** The first `n` triples genTriples(d) would emit — a delete phase only
 * needs a batch-sized slice of quads that actually exist in the store. */
export function genTriplesPrefix(d: number, n: number): Quad[] {
    const out: Quad[] = [];
    for (let s = 0; s < d && out.length < n; s++)
        for (let p = 0; p < d && out.length < n; p++)
            for (let o = 0; o < d && out.length < n; o++) out.push(df.quad(nn(s), nn(p), nn(o)));
    return out;
}

File: python/bench/datasets.py
"""Dataset generation for the Python comparative benchmark.

A deliberate port of `js/bench/datasets.ts`, kept algorithmically identical --
same moduli, same term spellings, same probe selection -- so a row on the
Python dashboard tab measures the *same data* as the corresponding row on the
JavaScript tab. Any drift here silently makes the two tabs incomparable, so the
uniqueness argument and the cardinality defaults are reproduced verbatim rather
than re-derived.

Two things differ from the JS original, both forced by what the Python
libraries accept:

  * The output is a serialized N-Triples/N-Quads *file*, not an array of term
    objects. Every library under test ingests from a file (`serialize_rdf`,
    `bulk_load`, `rdf2cottas`, `Graph.parse`), and lightrdf has no store at all
    -- a file is the only input all five share.
  * Generation streams to disk. Materializing 2M quads as Python objects first
    would cost more memory than any store under test, and this module runs in
    the same process as the adapter whose peak RSS is being attributed.

PURITY CONTRACT: standard library only. This module is imported by every
adapter's isolated virtualenv, which by construction does not have the other
adapters' dependencies installed.
"""

from __future__ import annotations

import os
from dataclasses import dataclass
from math import gcd, log
from typing import Iterator, Optional

BASE = "http://data.example.org"


# ─── Cardinality-controlled dataset ──────────────────────────────────────────
#
# Term cardinality is an explicit knob, independent of row count: drawing D**3
# rows from a namespace of D IRIs would make every store's term handling --
# dictionaries, interning, string storage -- invisible to the benchmark.
#
# Uniqueness: term indices are `i % k` per role, so quad i maps to the residue
# tuple (i mod nSubj, i mod nPred, i mod nObj, i mod nGraph). By the Chinese
# Remainder Theorem that map is injective over `i < lcm(...)`, so making the
# four moduli pairwise coprime (their lcm is then their product) and checking
# the product covers `n` guarantees every generated quad is distinct, with no
# dedupe set -- which at these row counts would itself dominate memory.
#
# Deliberate consequence: index 0 exists for every role, so row 0 satisfies all
# four single-term probes and every pattern below matches at least one row.

SUBJECT_RATIO_DEFAULT = 0.1


@dataclass(frozen=True)
class DatasetOpts:
    """Distinct terms per role. `subject_ratio` is the reciprocal of how many
    triples describe each resource -- 0.1 is ten triples per subject, which is
    what makes the dictionary a meaningful fraction of the dataset."""

    subject_ratio: float = SUBJECT_RATIO_DEFAULT
    predicates: int = 32
    object_ratio: float = 0.5
    literal_frac: float = 0.4
    graphs: int = 1


def dataset_opts(**overrides) -> DatasetOpts:
    """Env-overridable defaults, so a sweep can vary cardinality without code."""
    base = DatasetOpts(
        subject_ratio=float(os.environ.get("BENCH_SUBJ_RATIO", SUBJECT_RATIO_DEFAULT)),
        predicates=int(os.environ.get("BENCH_PREDICATES", 32)),
        object_ratio=float(os.environ.get("BENCH_OBJ_RATIO", 0.5)),
        literal_frac=float(os.environ.get("BENCH_LITERAL_FRAC", 0.4)),
        graphs=int(os.environ.get("BENCH_GRAPHS", 1)),
    )
    return DatasetOpts(**{**base.__dict__, **{k: v for k, v in overrides.items() if v is not None}})


@dataclass(frozen=True)
class Moduli:
    n_subj: int
    n_pred: int
    n_obj: int
    n_graph: int
    terms: int


def moduli(n: int, opts: Optional[DatasetOpts] = None) -> Moduli:
    """Per-role term counts: the requested values nudged up until pairwise coprime."""
    o = opts or dataset_opts()
    want = [
        max(1, round(n * o.subject_ratio)),
        max(1, round(o.predicates)),
        max(1, round(n * o.object_ratio)),
        max(1, round(o.graphs)),
    ]
    got: list[int] = []
    for w in want:
        k = w
        while any(gcd(g, k) != 1 for g in got):
            k += 1
        got.append(k)
    n_subj, n_pred, n_obj, n_graph = got

    # Product == lcm because they are pairwise coprime; see the CRT note above.
    # Use logs: the product overflows long before it matters numerically.
    if sum(log(k) for k in got) < log(n):
        raise ValueError(
            f"dataset cardinality too low for {n} distinct quads: "
            f"{n_subj}x{n_pred}x{n_obj}x{n_graph} cannot cover it -- raise a ratio"
        )
    # The default graph is a term too, and it is not one of the named graphs.
    terms = n_subj + n_pred + n_obj + (1 if n_graph == 1 else n_graph)
    return Moduli(n_subj, n_pred, n_obj, n_graph, terms)


# ─── Term spellings (must match js/bench/datasets.ts exactly) ────────────────

def subject_term(i: int) -> str:
    return f"<{BASE}/resource/2026/subject/{i:09d}>"


def predicate_term(i: int) -> str:
    return f"<{BASE}/ontology/2026/property/{i:04d}>"


def graph_term(i: int) -> str:
    return f"<{BASE}/graph/2026/named/{i:06d}>"


def object_term(i: int, literal_frac: float) -> str:
    """Objects alternate IRI/literal deterministically, in `literal_frac` proportion.

    The literal body is drawn from the same fixed alphabet as the JS generator
    and contains no quote, backslash, or newline, so N-Triples quoting here is
    a bare wrap rather than a general escaper -- keep it that way, or escaping
    cost lands in every adapter's parse timing.
    """
    if i % 10 < round(literal_frac * 10):
        return f'"descriptive object value number {i:09d}"'
    return f"<{BASE}/resource/2026/object/{i:09d}>"


def quad_at(i: int, m: Moduli, literal_frac: float) -> tuple[str, str, str, Optional[str]]:
    """Quad `i` of the dataset, as N-Triples-ready term strings. Pure in `i`, so
    probes and prefixes derive without materializing the dataset."""
    return (
        subject_term(i % m.n_subj),
        predicate_term(i % m.n_pred),
        object_term(i % m.n_obj, literal_frac),
        None if m.n_graph == 1 else graph_term(i % m.n_graph),
    )


def iter_quads(n: int, opts: Optional[DatasetOpts] = None) -> Iterator[tuple[str, str, str, Optional[str]]]:
    o = opts or dataset_opts()
    m = moduli(n, o)
    for i in range(n):
        yield quad_at(i, m, o.literal_frac)


def write_dataset(path: str, n: int, opts: Optional[DatasetOpts] = None) -> str:
    """Stream `n` distinct quads to an N-Triples (or N-Quads) file at `path`.

    Written once by the orchestrator and shared by every adapter process: the
    file is the common input, so no adapter pays for another's generation, and
    generation cost never lands inside a measured region.
    """
    o = opts or dataset_opts()
    tmp = path + ".partial"
    with open(tmp, "w", encoding="utf-8", buffering=1 << 20) as f:
        for s, p, ob, g in iter_quads(n, o):
            f.write(f"{s} {p} {ob} .\n" if g is None else f"{s} {p} {ob} {g} .\n")
    os.replace(tmp, path)  # never leave a half-written dataset that looks complete
    return path


def fresh_quads(n: int) -> list[tuple[str, str, str]]:
    """A batch of quads in a namespace disjoint from `write_dataset`'s, so the
    add phase inserts rows the store cannot already hold."""
    return [
        (
            f"<{BASE}/fresh/2026/subject/{i:09d}>",
            f"<{BASE}/fresh/2026/property/0000>",
            f"<{BASE}/fresh/2026/object/{i:09d}>",
        )
        for i in range(n)
    ]


def dataset_prefix(n: int, take: int, opts: Optional[DatasetOpts] = None) -> list[tuple[str, str, str, Optional[str]]]:
    """The first `take` quads `write_dataset(n)` would produce. The delete phase
    needs a batch-sized slice of quads that actually exist in the store."""
    o = opts or dataset_opts()
    m = moduli(n, o)
    return [quad_at(i, m, o.literal_frac) for i in range(min(take, n))]


# ─── Query patterns ─────────────────────────────────────────────────────────
#
# Probe terms are fixed at index 0, so every probe matches row 0 at minimum and
# no pattern can silently measure a zero-row query. Selectivity follows the
# data: at the defaults `S` matches ~10 rows, `P` ~n/32, `O` ~2, and the
# conjunctions narrow to a handful. Each worker records the matched-row count
# alongside the timing so the figures stay interpretable.


@dataclass(frozen=True)
class Pat:
    name: str
    s: Optional[str]
    p: Optional[str]
    o: Optional[str]
    g: Optional[str]


FULL_SCAN_PATTERN = Pat("full", None, None, None, None)


def dataset_probes(n: int, opts: Optional[DatasetOpts] = None) -> dict[str, list[Pat]]:
    o = opts or dataset_opts()
    m = moduli(n, o)
    s0 = subject_term(0)
    p0 = predicate_term(0)
    o0 = object_term(0, o.literal_frac)
    g0 = None if m.n_graph == 1 else graph_term(0)
    # A single-graph dataset has no graph term to bind, which would quietly turn
    # `G` into a second full scan and `SPOG` into `SPO` -- two probes measuring
    # something other than what their names claim. Return no quad probes instead,
    # so a caller that forgot to pass the quads options gets nothing rather than
    # plausible-looking wrong rows.
    quads = (
        []
        if g0 is None
        else [Pat("G", None, None, None, g0), Pat("SPOG", s0, p0, o0, g0)]
    )
    return {
        "triples": [
            Pat("S", s0, None, None, None),
            Pat("P", None, p0, None, None),
            Pat("O", None, None, o0, None),
            Pat("PO", None, p0, o0, None),
            Pat("SPO", s0, p0, o0, None),
        ],
        "quads": quads,
        "full": [FULL_SCAN_PATTERN],
    }


# ─── Dense-cube datasets (the instrumented suite's routing-pattern data) ─────
#
# A d**3 / d**4 cube over a tiny closed namespace: every routing decision is
# exercised while the dataset stays small enough to instrument. Deliberately
# NOT cardinality-realistic -- 3*d distinct terms over d**3 rows is exactly the
# pathology the generator above exists to avoid -- so term-handling-sensitive
# tasks must use `write_dataset` instead; the cube is for tasks where routing,
# not term volume, is what is being measured.
#
# Ported from `js/bench/datasets.ts` alongside the generator above, so the
# instrumented Python and JavaScript suites measure the same shapes.

EX = "http://example.org/#"

XSD_INTEGER = "<http://www.w3.org/2001/XMLSchema#integer>"


def nn(n) -> str:
    return f"<{EX}{n}>"


def escape_literal(value: str) -> str:
    """N-Triples escaping for a literal's lexical form.

    Needed only by `write_literal_triples`, whose escape-bearing case carries
    quotes, backslashes and newlines on purpose; every other generator here
    emits values that pass through unchanged.
    """
    out = value.replace("\\", "\\\\").replace('"', '\\"')
    return out.replace("\n", "\\n").replace("\r", "\\r").replace("\t", "\\t")


def write_triples_cube(path: str, d: int) -> str:
    """d**3 triples in the default graph: (ex#s, ex#p, ex#o)."""
    with open(path, "w", encoding="utf-8", buffering=1 << 20) as f:
        for s in range(d):
            for p in range(d):
                for o in range(d):
                    f.write(f"{nn(s)} {nn(p)} {nn(o)} .\n")
    return path


def write_quads_cube(path: str, d: int) -> str:
    """d**4 quads across named graphs: (ex#s, ex#p, ex#o, ex#g)."""
    with open(path, "w", encoding="utf-8", buffering=1 << 20) as f:
        for s in range(d):
            for p in range(d):
                for o in range(d):
                    for g in range(d):
                        f.write(f"{nn(s)} {nn(p)} {nn(o)} {nn(g)} .\n")
    return path


def write_literal_triples(path: str, d: int) -> str:
    """d**3 triples whose objects are literals -- plain, language-tagged, typed,
    and escape-bearing (quotes, backslashes, newlines, and `"@`/`^^` lookalikes
    inside the value). The other cube datasets are named-node-only, so tasks
    over this one are the only cube coverage of literal escaping."""
    i = 0
    with open(path, "w", encoding="utf-8", buffering=1 << 20) as f:
        for s in range(d):
            for p in range(d):
                for o in range(d):
                    if i % 4 == 0:
                        obj = f'"plain value {o}"'
                    elif i % 4 == 1:
                        obj = f'"tagged value {o}"@en'
                    elif i % 4 == 2:
                        obj = f'"{o}"^^{XSD_INTEGER}'
                    else:
                        raw = f'say "hi"@home ^^ line\nbreak back\\slash {o}'
                        obj = f'"{escape_literal(raw)}"'
                    f.write(f"{nn(s)} {nn(p)} {obj} .\n")
                    i += 1
    return path
